### ISSUE TITLE
UnboundLocalError: local variable 'w' referenced before assignment for JPG

### ISSUE THREAD

**Issue by ahuang11** [holoviz/panel#1959]
```
class JPG(ImageBase):

    imgtype = 'jpg'

    @classmethod
    def _imgshape(cls, data):
        import struct
        b = BytesIO(data)
        b.read(2)
        c = b.read(1)
        while (c and ord(c) != 0xDA):
            while (ord(c) != 0xFF): c = b.read(1)
            while (ord(c) == 0xFF): c = b.read(1)
            if (ord(c) >= 0xC0 and ord(c) <= 0xC3):
                b.read(3)
                h, w = struct.unpack(">HH", b.read(4))
                break
            else:
                b.read(int(struct.unpack(">H", b.read(2))[0])-2)
            c = b.read(1)
        return int(w), int(h)
```

```

~/anaconda3/envs/py3/lib/python3.8/site-packages/panel/pane/image.py in _get_properties(self)
    105         if not isinstance(data, bytes):
    106             data = base64.b64decode(data)
--> 107         width, height = self._imgshape(data)
    108         if self.width is not None:
    109             if self.height is None:

~/anaconda3/envs/py3/lib/python3.8/site-packages/panel/pane/image.py in _imgshape(cls, data)
    187                 b.read(int(struct.unpack(">H", b.read(2))[0])-2)
    188             c = b.read(1)
--> 189         return int(w), int(h)
    190 
    191 

UnboundLocalError: local variable 'w' referenced before assignment
```

**Comment by ahuang11** [comment]
BytesIO object.

**Comment by ahuang11** [comment]
Oops, I was using JPG for PNG images

**Comment by jbednar** [comment]
In general you'll get this error for anything that's not a valid JPG image, so I'm reopening this as a request for a more useful error message to convey that.

**Comment by philippjfr** [comment]
Tbh, I think I'll create a generic `Image` pane which wraps around the other specific panes and has higher precedence. I've been annoyed before that I could not update an existing PNG pane with a JPG for instance and since they are rendered using the same underlying model there's no reason not to do this.

**Comment by jbednar** [comment]
Yes, I had the same thought; no need to care about the specific type of image it is.

**Comment by linehammer** [comment]
Python doesn't have variable declarations , so it has to figure out the scope of variables itself. It does so by a simple rule: If there is an assignment to a variable inside a function, that variable is considered local .  In most cases "local variable [referenced before assignment](http://net-informations.com/python/err/local.htm)"  will occur when trying to modify a local variable before it is actually assigned within the local scope. Python doesn't have variable declarations, so it has to figure out the scope of variables itself. It does so by a simple rule: If there is an assignment to a variable inside a function, that variable is considered local. 

Python has lexical scoping by default, which means that although an enclosed scope can access values in its enclosing scope, it cannot modify them (unless they're declared global with the global keyword). All variable assignments in a function store the value in the local symbol table; whereas variable references first look in the local symbol table, then in the global symbol table, and then in the table of built-in names. Thus, global variables cannot be directly assigned a value within a function (unless named in a global statement), although they may be referenced.
